### PR TITLE
Model residual maps

### PR DESCRIPTION
**Pull request by cbrisboi**
This branch should now properly implement making model and residuals maps. 
I also have added a script to convert the hdf5 files to fits for compatibility with older hawc tools (this could probably use some better info from the origin maptrees, but it approximately works so far.

### CODE REVIEW

File: hawc_hal/healpix_handling/sparse_healpix.py [hunk 1]
```diff
@@ -74,6 +73,38 @@ def __init__(self, partial_map, pixels_ids, nside, fill_value=UNSEEN):
 
         super(SparseHealpix, self).__init__(sparse=True, nside=nside)
 
+    def __add__(self, other_map):
+
+        # Make sure they have the same pixels
+        assert np.array_equal(self.pixels, other_map.pixels)
+
+        # inflate them
```
Comment (giacomov) [inline hawc_hal/healpix_handling/sparse_healpix.py]: There is no need to inflate them. You can just use `.as_partial` which returns a numpy array with the same shape as the selected pixels. So for example to do what you want to do you can do:

```python
# add them
partial_added = self.as_partial() + other_map.as_partial()
sparse_added = SparseHealpix(dense_added, self.pixels, self.nside)
```

File: hawc_hal/healpix_handling/sparse_healpix.py [hunk 2]
```diff
@@ -74,6 +73,38 @@ def __init__(self, partial_map, pixels_ids, nside, fill_value=UNSEEN):
...
+        
+        # add them
+        dense_added = big_self_map + big_other_map
+
+        # deflate
+        sparse_added = SparseHealpix(dense_added[self.pixels], self.pixels, self.nside)
+        
+        return sparse_added
+
+    def __sub__(self, other_map):
```
Comment (giacomov) [inline hawc_hal/healpix_handling/sparse_healpix.py:87]: same comment as for the addition

File: hawc_hal/healpix_handling/sparse_healpix.py [hunk 3]
```diff
@@ -86,7 +117,7 @@ def as_dense(self):
         new_map = np.full(self.npix, self._fill_value)
 
         # Assign the active pixels their values
-        new_map[self._pixels_ids] = self._partial_map
+        new_map[self.pixels] = self._partial_map
```
Comment (giacomov) [inline hawc_hal/healpix_handling/sparse_healpix.py]: within a method of the class you can access private members (i.e., members starting with `_`). It is actually faster, because using properties have a slight overhead (although negligible most of the time).

Please change this line back to what it was:

`new_map[self._pixels_ids] = self._partial_map`

File: hawc_hal/healpix_handling/sparse_healpix.py [hunk 4]
```diff
@@ -100,6 +131,10 @@ def set_new_values(self, new_values):
 
         self._partial_map[:] = new_values
 
+    @property
+    def pixels(self):
```
Comment (giacomov) [inline hawc_hal/healpix_handling/sparse_healpix.py]: I would call this `pixel_ids`, to make clearer that these are the Healpix IDs of the pixels, not their values

File: hawc_hal/maptree/__init__.py [hunk 5]
```diff
@@ -1 +1,3 @@
-from map_tree import map_tree_factory
\ No newline at end of file
+from map_tree import map_tree_factory
+from map_tree import MapTree
```
Comment (giacomov) [inline hawc_hal/maptree/__init__.py]: What is the need of adding these here?

Adding an import on a `__init__` means that the user will have access to those classes. Since the user should not create a `MapTree` directly, nor a `DataAnalysisBin` directly (these are for internal use), we should not import them into this `__init__`.

If you need to have access to those from another subdirectory, you can just say:

`from hawc_hal.maptree.map_tree import MapTree`

for example.

**Review by giacomov — CHANGES_REQUESTED** [review]
Thank you very much for this nice addition!
I left some comments with some minor modifications required.

**Comment by cbrisboi** [comment]
I made a commit to reflect your feedback, before realizing that its much easier to read on github then via email. After discussing with @henrikef I also made a model map(but did not commit it) which might be used to produce some unit tests using the already provided detector response file in hal.

**Comment by colasri** [comment]
There is a map already in `pa-pub.umd.edu:/data/disk01/home/giacomov/test_data`, which is used in the HAL test:

```
# Test the HAWC plugin
# First copy the data in /data/disk01/home/giacomov/test_data at UMD to
# the directory test_data
mkdir hawc_test_data
rsync --progress 'pa-pub.umd.edu:/data/disk01/home/giacomov/test_data/*' hawc_test_data
export HAWC_3ML_TEST_DATA_DIR=$SOFTWARE_BASE/3ML_tests/hawc_test_data
pytest -vv --pyargs threeML -k "hawc"
```
Could you use the same map for your test, or do you really need a new file?

**Comment by cbrisboi** [comment]
Colas - I don't as yet have a test. It was something @henrikef told me I should perhaps look into for the future.

I was talking to @henrikef earlier today, she suggested I generate the hd5 file for the test directly, and so that would only require adding the script, rather than a whole new map.  The idea would be to check that the model/residual maps were generated properly, since I haven't looked at those maps I cant say for sure. I don't have a preference per se, but generating the needed maptree on the fly might be a good plan for bandwidth and things. I'm not sure what the best plan would be.

**Review by giacomov — APPROVED** [review]

**Comment by giacomov** [comment]
In general a test should accomplish two things:

* Bare functionality: run through all the pieces of new code to test whether they execute without errors or not

* Correctness: check that the results of a real use case are correct.

While for the first point any data is good, for the second one you need a dataset that yields a predictable answer, so you can check that the answer is what you expect. I have no preference on whether to use or not a new map. In the latter case, please do make it hd5 and with a fairly small ROI (< 20 deg) so that the size of the file is small.

I will wait for the implementation of the tests before merging this in.

**Comment by cbrisboi** [comment]
My idea for a test is to generate a model map using the geminga_maptree for background and the detector response file we have in hal already. Thats slightly more portable than adding a whole map. I'm thinking to just read in the file and compare that the model that was saved is the same as what is read back in. Does that sound reasonable?

**Comment by giacomov** [comment]
Yes, it does! Make sure to cover in the test also corner cases and exceptions. Have a look at the other tests for inspiration, and feel free to ask questions either here or on Slack.

**Comment by giacomov** [comment]
@cbrisboi I am looking at your test so far. I understand that it is work in progress, but I thought I can give you some more pointers to avoid too much work on your side.

* First, tests *must* be within functions called test_[something]

* Second, look into fixtures: https://docs.pytest.org/en/latest/fixture.html

Now you can understand that all the code you have up until line 33 can be avoided.

You can get exactly the same by doing:

```python
def test_model_residual_maps(geminga_roi, geminga_maptree, geminga_response):
    
    hawc = HAL("HAWC", geminga_maptree, geminga_response, geminga_roi)
    
    # Now you can continue as before
    hawc.set_active_measurements(1, 9)
    
    # Here you can put the rest of your code
    ...
```

**Comment by cbrisboi** [comment]
@giacomov I'm not sure that I fully addressed the your statement that the code up to line 33 can be avoided, but I hope what I did commit is clear. The test is in a function now, and uses a fixture as input.
